tximeta in RNA-seq - Part 1: updating 02-gastric_cancer_tximeta.Rmd (#343)
AlexsLemonade/training-modules: RNA-seq/02-gastric_cancer_tximeta.Rmd

---

> title: "Gastric cancer: gene-level summarization with `tximport`"

Reviewer: I am wondering if we should include `tximeta` in the title as well, I felt strongly that we should at first glance, to stay consistent with the file name, but now I feel less strongly.

Author: Yeah I haven't been 100% sure about when to reference `tximeta` and when to reference `tximport`.

For more information about `tximeta`, see [this excellent vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/tximeta/inst/doc/tximeta.html) from Love, Soneson, Hickey, and Patro.

We'll need the `quant.sf` files for all the samples in an experiment.

## Libraries and functions

```{r library, live = TRUE}
# Load the tximeta package
library(tximeta)

# Load the SummarizedExperiment package
library(SummarizedExperiment)
```

## Directories and files

```{r directories, live = TRUE}
# directory where the quant files are located, each sample is its own
# directory
quant_dir <- file.path("data", "gastric-cancer", "salmon_quant")

# create a directory to hold the tximport results if it doesn't exist yet
txi_dir <- file.path("data", "gastric-cancer", "tximport")
if (!dir.exists(txi_dir)) {
  dir.create(txi_dir, recursive = TRUE)
}
```

```{r input-names}
# the quant files themselves
sf_files <- list.files(quant_dir, recursive = TRUE, full.names = TRUE,
                       pattern = "quant.sf")
```

**Output**

```{r output-names, live = TRUE}
# Name the output gastric-cancer_tximport.RDS and use the directory created 
# above as the rest of the path

> txi_out_file <- file.path(txi_dir, "gastric-cancer_tximport.RDS")

Reviewer: Based on my interpretation @jashapiro's [comment here](https://github.com/AlexsLemonade/training-modules/pull/343#issuecomment-780807162)
> I think we can keep tximport in at least some of these places as long as we clearly lay out that the txi in tximeta stands for tximport (and the tx stands for transcript expression?). So the diagrams should be able to stay, and probably the directory names, but since the object structure is different we should probably change the .rds filenames

I believe we'll want to replace the `tximport` with `tximeta` here in the file name of the output RDS (please forgive me if I misinterpreted this comment!)

Author: I think you are right.

This is exactly what `stringr::word` allows us to do: split up the file paths
by `/` and extract the sample names.

```{r sample_names}
sample_names <- stringr::word(sf_files, 4, sep = "/")
sample_names
```

`tximeta` needs a data.frame with these columns: 
- a `files` column  with the file paths to the quant.sf files
- a `names` column with the sample names

```{r names_sf_files}
coldata <- data.frame(files = sf_files, 
                      names = sample_names, 
                      stringsAsFactors = FALSE)
```

## Set up our tximeta object

Using our `coldata` data frame we set up, we can run the `tximeta()` which will 
extract species specific information.

```{r read_tsv, live = TRUE}
txi_meta <- tximeta(coldata)
```

*There are extra steps needed if you are using any species other than human or mouse.
See our other species examples in `tximeta_for_other_species.Rmd`. 

### Summarize to gene

We'll summarize to the gene level using the `summarizeToGene()` function.

```{r make-txi}
# Summarize to the gene-level
gene_summarized <- summarizeToGene(txi_meta) 
```

We can use the `class` function to see what type of object `gene_summarized` is.

```{r class-txi, live = TRUE}

> # Check what type of object gene_summarized is

Reviewer: ```suggestion
# Check what type of object `gene_summarized` is
```

Reviewer: I think we'll still want the backticks around object names in comments?

Author: I think so.

We more specifically have a `RangedSummarizedExperiment` which is a more specific type of `SummarizedExperiment`.

`SummarizedExperiment` objects have this general structure:  

![SummarizedExperiment](diagrams/SummarizeExperiment-structure.png)

This figure is from this handy vignette about [`SummarizedExperiment` objects](https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

The `assay` slot in `SummarizedExperiment`s hold the experiment data, a gene x sample matrix. 
In this case, it has our gene-level abundance information. 

Multiple `assays` can be stored in an `SummarizedExperiment` and we can use the `assayNames()` function to see what assays are included in `gene_summarized`. 

```{r assay-names, live = TRUE}
assayNames(gene_summarized)
```

If we want to extract an `assay`'s data, we can use `assay()` function and 
specify the name of the assay we want to extract.

```{r assay, live = TRUE}
counts_mat <- assay(gene_summarized, "counts")
```

> We can use the `class` function to see what type of object the `tximport`

Reviewer: ```suggestion
We can use the `class` function to see what type of object the `assay()`
```

I believe the `tximport()` function is no longer implemented in this notebook, if I am not mistaken. Did you mean the `assay()` function here?

Author: Yes. I missed that this comment needed updating.

function returns.

```{r counts, live = TRUE}

> # Check what type of object counts_mat is

Reviewer: ```suggestion
# Check what type of object `counts_mat` is
```

class(counts_mat)
```

### Save to file

We could use `readr::write_tsv` to save `counts_mat` only but `gene_summarized` 
has a lot of information stored here beyond the counts, so we may want to save 
all of this to a RDS object. 

```{r write-txi, live = TRUE}

> # Write gene_summarized to RDS object

Reviewer: ```suggestion
# Write `gene_summarized` to RDS object
```

```{r sf_files, live = TRUE}
# Let's look at the full path for the quant.sf files
sf_files
```

Let's extract the _sample_ names from the **file paths** using the `stringr` 
package.

Notice how the file path is separated by `/`.
If we were to split up this character by `/`, the fourth item is the sample 
name.
This is exactly what `stringr::word` allows us to do: split up the file paths
by `/` and extract the sample names.

```{r sample_names}
sample_names <- stringr::word(sf_files, 4, sep = "/")
sample_names
```

`tximeta` needs a data.frame with these columns: 
- a `files` column  with the file paths to the quant.sf files
- a `names` column with the sample names

```{r names_sf_files}
coldata <- data.frame(files = sf_files, 
                      names = sample_names, 
                      stringsAsFactors = FALSE)
```

## Set up our tximeta object

Using our `coldata` data frame we set up, we can run the `tximeta()` which will 
extract species specific information.

```{r read_tsv, live = TRUE}
txi_meta <- tximeta(coldata)
```

*There are extra steps needed if you are using any species other than human and mouse.

> See our other species examples in `tximeta_for_other_species.Rmd`. 

Reviewer 2: I think until we have this, we should say something more oblique... Perhaps more along the lines of "`tximeta` currently works easily for most human and mouse datasets, but requires a few more steps for other species." I would point to the vignette until then: https://bioconductor.org/packages/release/bioc/vignettes/tximeta/inst/doc/tximeta.html#What_if_checksum_isn%E2%80%99t_known

---
title: "Gastric cancer: gene-level summarization with `tximport`"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

**CCDL 2021**

In this notebook, we'll import transcript level output from `salmon quant` and

> summarize it to the gene level using 'transcript expression import' - [`tximport`](https://bioconductor.org/packages/release/bioc/html/tximport.html) and 'tximport meta' - [`tximeta`](https://bioconductor.org/packages/release/bioc/html/tximeta.html). 

Author: > Only making this comment here, though it applies throughout. I think we were starting to standardize on one sentence per line, for easier tracking of revisions & more consistent line numbers. (If we were to do more aggressive line breaks, this would need to be broken up, but I would just make this one line.)

My impulse was to do this but then I thought I remembered our standard for this repo was different. Is this something that can be added to `CONTRIBUTING.md`? Also I think the spacing/line stuff is weird in a lot of places, so if there's a way we can fix it all at once that's probably good. I'll change it in this notebook to start.

> summarize it to the gene level using 'transcript expression import' - [`tximport`](https://bioconductor.org/packages/release/bioc/html/tximport.html) and 'tximport meta' - [`tximeta`](https://bioconductor.org/packages/release/bioc/html/tximeta.html). 

Reviewer 2: I think it was different for a while, but we were starting to standardize as notebooks got modified. Pretty sure we changed everything in scRNA-seq, but possibly not.

![](diagrams/rna-seq_5.png)

For more information about `tximeta`, see [this excellent vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/tximeta/inst/doc/tximeta.html) from Love, Soneson, Hickey, and Patro.

We'll need the `quant.sf` files for all the samples in an experiment.

## Libraries and functions

> ```{r library, live = TRUE}

Reviewer 2: ````suggestion
```{r library, live = TRUE}
# Load magrittr for the pipe
library(magrittr)

````

# Load the tximeta package
library(tximeta)

# Load the SummarizedExperiment package
library(SummarizedExperiment)
```

## Directories and files

> 

Author: I put this type of comment right below here right before we use `list.files()`.

```

**Output**

```{r output-names, live = TRUE}
# Name the output gastric-cancer_tximport.RDS and use the directory created 
# above as the rest of the path
txi_out_file <- file.path(txi_dir, "gastric-cancer_tximeta.RDS")
```

### Naming

All output files from `salmon quant` we'll use with `tximeta` are named 
`quant.sf`.
The file names do not have any information about the sample name.

```{r sf_files, live = TRUE}
# Let's look at the full path for the quant.sf files
sf_files
```

Let's extract the _sample_ names from the **file paths** using the `stringr` 
package.

Notice how the file path is separated by `/`.
If we were to split up this character by `/`, the fourth item is the sample 
name.
This is exactly what `stringr::word` allows us to do: split up the file paths
by `/` and extract the sample names.

```{r sample_names}
sample_names <- stringr::word(sf_files, 4, sep = "/")
sample_names
```

`tximeta` needs a data.frame with at least these two columns: 
- a `files` column  with the file paths to the quant.sf files
- a `names` column with the sample names

```{r names_sf_files}
coldata <- data.frame(files = sf_files, 
                      names = sample_names, 

>                       stringsAsFactors = FALSE)

Author: Oh yeah! :tada:

```

> ### Set up metadata

Author: Should probably be H2?

Reviewer 2: Sure, but maybe bring it up to above the creation of the `coldata` data frame? I would consider all of this making the metadat that will go into the `tximeta` call.

And now we'll read in the sample metadata.

```{r sample_meta_df, live = TRUE}
# Read in the sample metadata TSV file and have a look
sample_meta_df <- readr::read_tsv(meta_file)
sample_meta_df
```

> We'll want this information to be added to the `coldata`. 

Reviewer 2: I don't think we will have covered joins at this point, so we can either follow the previous model of checking that samples are in the same order and appending, or we probably need a more extensive description of what is happening here.

What is the metadata that we have? How is it arranged? (We have them look, but we should discuss the content, at least with a questioning prompt)

Author: We cover joins in intro-to-r: https://github.com/AlexsLemonade/training-modules/blob/ebb928ed6499cc7ae41854e470998f75c4757fe4/intro-to-R-tidyverse/03-intro_to_tidyverse.Rmd#L493

Reviewer 2: Yay past us!

```{r join-sample_meta_df}
coldata <- coldata %>%
  dplyr::inner_join(sample_meta_df, by = c("names" = "srr_accession"))

coldata
```

> ## Set up our tximeta object

Reviewer 2: Something that is less about the object than the action?
```suggestion
## Import expression data with `tximeta`
```

> Using our `coldata` data frame we set up, we can run the `tximeta()` which will 
> extract species specific information.

Reviewer 2: More context!
```suggestion
Using the `coldata` data frame that we set up, we can now run the `tximeta()` to import our expression data while automatically finding and associating the transcript annotations that were used when we performed the quantification.
```

> ```{r read_tsv, live = TRUE}

Reviewer 2: update chunk name

> txi_meta <- tximeta(coldata)

Reviewer 2: The first time this runs, we will get a message about setting up a cache: 
> tximeta needs a BiocFileCache directory to access and save TxDb objects.
Do you wish to use the default directory: '/home/ccdl/josh/.cache/BiocFileCache'?
If not, a temporary directory that is specific to this R session will be used.

> You can always change this directory later by running: setTximetaBFC()
Or enter [0] to exit and set this directory manually now.
This location can also be set by environmental variable TXIMETA_HUB_CACHE. 

We should make some note about what people should do (use the cache, but it won't hurt much if you don't... just slower next time.)

Reviewer 2: I also realized that this may cause trouble with automation... we may need to add something to make-live.R or the github actions to stop it from prompting for this, but maybe they were smart enough to detect non-interactive sessions. This is a later problem.

> txi_meta <- tximeta(coldata)

Reviewer 2: I don't love naming this `txi_meta`, as I think it is potentially confusing. Maybe `txi_data`?

```

*There are extra steps needed if you are using any species other than human or mouse.
See our other species examples in `tximeta_for_other_species.Rmd`. 

### Summarize to gene

We'll summarize to the gene level using the `summarizeToGene()` function.

> ```{r make-txi}

Reviewer 2: update chunk name

> # Summarize to the gene-level

Reviewer 2: ```suggestion
# Summarize to the gene level
```

gene_summarized <- summarizeToGene(txi_meta) 
```

We can use the `class` function to see what type of object `gene_summarized` is.

```{r class-txi, live = TRUE}
# Check what type of object `gene_summarized` is
class(gene_summarized)
```

This tells us that `gene_summarized` is an object called a [`SummarizedExperiment`](https://www.rdocumentation.org/packages/SummarizedExperiment/versions/1.2.3/topics/SummarizedExperiment-class) which can be handled by functions from the package of the same name. 
We more specifically have a `RangedSummarizedExperiment` which is a more specific type of `SummarizedExperiment`.

`SummarizedExperiment` objects have this general structure:  

![SummarizedExperiment](diagrams/SummarizeExperiment-structure.png)

This figure is from this handy vignette about [`SummarizedExperiment` objects](https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

The `assay` slot in `SummarizedExperiment`s hold the experiment data, a gene x sample matrix. 

> In this case, it has our gene-level abundance information. 

Reviewer 2: I think it is worth calling out the row data and column data at this point as well. People will be happy to see that `rowData(gene_summarized)` has more than just ENSG IDs; it has symbols too!

Multiple `assays` can be stored in an `SummarizedExperiment` and we can use the `assayNames()` function to see what assays are included in `gene_summarized`. 

```{r assay-names, live = TRUE}

> assayNames(gene_summarized)

Reviewer 2: Somebody is going to ask what the `"abundance"` assay is, we should probably note that it is TPM (I wish this were easier information to surface: I expected this was the case, but struggled to find confirmation. I finally found it with `?"tximeta-package"`, but it wasn't obvious in function descriptions!)

Author: I had this in the next notebook and then realized it should be here so I switched it in the next PR but apparently not here. Whoops

`tximeta` eases some of the burden of import by automatically identifying the correct set of annotation data to append to many data sets ([Love _et al._ 2020](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007664)).

![](diagrams/rna-seq_5.png)

For more information about `tximeta`, see [this excellent vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/tximeta/inst/doc/tximeta.html) from Love _et al_.

## Libraries and functions

```{r library}
# Load magrittr for the pipe
library(magrittr)

# Load the tximeta package
library(tximeta)

# Load the SummarizedExperiment package
library(SummarizedExperiment)
```

## Directories and files

```{r directories, live = TRUE}
# directory where the data are located
data_dir <- file.path("data", "gastric-cancer")

# directory where the quant files are located, each sample is its own
# directory
quant_dir <- file.path(data_dir, "salmon_quant")

# create a directory to hold the tximport results if it doesn't exist yet
txi_dir <- file.path(data_dir, "tximport")
if (!dir.exists(txi_dir)) {
  dir.create(txi_dir, recursive = TRUE)
}
```

> We'll need the `quant.sf` files for all the samples in an experiment which we have stored in `quant_dir`. 

Reviewer 2: We should explain here what these files are, including not only that this is the output from `salmon`, but also a bit of a mention of the file structure. We might want to also make a note about what `list.files()` is doing here, as I don't think we have covered it before. 

This is also a place where we should start to address https://github.com/AlexsLemonade/training-modules/issues/360, so I am happy to leave this for another PR.

Author: Yeah, seems better suited for next PR.

```{r input-names}
# the quant files themselves
sf_files <- list.files(quant_dir, recursive = TRUE, full.names = TRUE,
                       pattern = "quant.sf")
```

```{r metadata-file}
# sample metadata file
meta_file <- file.path(data_dir, "gastric-cancer_metadata.tsv")
```

**Output**

```{r output-names, live = TRUE}
# Name the output gastric-cancer_tximport.RDS and use the directory created 
# above as the rest of the path
txi_out_file <- file.path(txi_dir, "gastric-cancer_tximeta.RDS")
```

## File names

All output files from `salmon quant` we'll use with `tximeta` are named `quant.sf`.
Unfortunately, this means that the file names themselves do not have any information about the sample they come from!

```{r sf_files, live = TRUE}
# Let's look at the full path for the quant.sf files
sf_files
```

Let's extract the _sample_ names from the **file paths** using the `stringr` package.

Notice how the file path is separated by `/`.

> If we were to split up this character string by `/`, the second to last item is the sample name.

Reviewer 2: I thought we could point out that the sample name here didn't just happen, we chose to use it for the directory ame
```suggestion
If we were to split up this character string by `/`, the second to last item is the sample names (because we used them as  directory names for the `salmon` output).
```

This is exactly what `stringr::word()` allows us to do: split up the file paths by `/` and extract the sample names.

```{r sample_names}
sample_names <- stringr::word(sf_files, -2, sep = "/")
sample_names
```

## Set up metadata

`tximeta` needs a data frame with at least these two columns: 
- a `files` column  with the file paths to the quant.sf files
- a `names` column with the sample names

```{r names_sf_files}
coldata <- data.frame(files = sf_files, 
                      names = sample_names)
```

We have more information about these samples stored in the metadata file that we will also want stored in `coldata`. 
Let's read in the sample metadata from the TSV file.

```{r sample_meta_df, live = TRUE}
# Read in the sample metadata TSV file and have a look
sample_meta_df <- readr::read_tsv(meta_file)
sample_meta_df
```

> We'll want this information to be added to the `coldata`. 

Reviewer 2: A bit more context here.
```suggestion
We'll want this information to be added to the `coldata`, which we can do by using a join function to match up the rows between the two data frames and combine them. 
```

```{r join-sample_meta_df}
coldata <- coldata %>%
  dplyr::inner_join(sample_meta_df, by = c("names" = "srr_accession"))

coldata
```

## Import expression data with `tximeta`

Using the `coldata` data frame that we set up, we can now run the `tximeta()` to import our expression data while automatically finding and associating the transcript annotations that were used when we performed the quantification.

> The first time you run `tximeta()` you will get a message about storing the data in a [cache](https://en.wikipedia.org/wiki/Cache_(computing) so that it will retrieve the data faster the next time.
> We recommend you use the cache.

Reviewer 2: I don't think the wikipedia link is very helpful, as it covers way more than a user needs to know here. 
```suggestion
The first time you run `tximeta()` you may get a message about storing downloaded transcriptome data in a cache directory so that it can retrieve the data more quickly the next time.
We recommend you use the cache, and accept the default location.
```

## Summarize to gene

We'll summarize to the gene level using the `summarizeToGene()` function.

```{r summarize-gene}
# Summarize to the gene level
gene_summarized <- summarizeToGene(txi_data) 
```

We can use the `class` function to see what type of object `gene_summarized` is.

```{r class, live = TRUE}
# Check what type of object `gene_summarized` is
class(gene_summarized)
```

This tells us that `gene_summarized` is an object called a [`SummarizedExperiment`](https://www.rdocumentation.org/packages/SummarizedExperiment/versions/1.2.3/topics/SummarizedExperiment-class) which can be handled by functions from the package of the same name. 
We more specifically have a `RangedSummarizedExperiment` which is a more specific type of `SummarizedExperiment`.

`SummarizedExperiment` objects have this general structure:  

![SummarizedExperiment](diagrams/SummarizeExperiment-structure.png)

This figure is from this handy vignette about [`SummarizedExperiment` objects](https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

> As shown in the diagram, we can use some of these functions to extract data from our `SummarizedExperiment` object. 

Reviewer 2: ```suggestion
As shown in the diagram, we can use some of the functions provided by the `SummarizedExperiment` package to extract data from our `gene_summarized` object. 
```

> `rowData()` on our object shows all the gene information that `tximeta` set up!

Reviewer 2: ```suggestion
For example, calling `rowData()` on our object shows all the gene information that `tximeta` set up!
```

```{r}
# rowData() shows us our gene annotation
rowData(gene_summarized)
```

> The `assay` slot in `SummarizedExperiment`s hold the experiment data, a gene x sample matrix. 
> In this case, it has our gene-level abundance information. 

Reviewer 2: Strictly speaking not all assays are gene x sample matrices (lengths is not) so I am adjusting this.
```suggestion
The `assay` slot in `SummarizedExperiment`s holds data from the experiment. 
In this case, it will include our gene-level abundance information stored as a gene x sample matrix. 
```

Author: Ah. Yeah. Good point.

This figure is from this handy vignette about [`SummarizedExperiment` objects](https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

As shown in the diagram, we can use some of the functions provided by the `SummarizedExperiment` package to extract data from our `gene_summarized` object. 
For example, calling `rowData()` on our object shows all the gene information that `tximeta` set up!

```{r}
# rowData() shows us our gene annotation
rowData(gene_summarized)
```

The `assay` slot in `SummarizedExperiment`s holds data from the experiment. 

> In this case, it will include our gene-level abundance information stored as a gene x sample matrix. 

Reviewer 2: Modifying to not create confusion with `abundance`
```suggestion
In this case, it will include our gene-level expression information stored as a gene x sample matrix. 
```
